src/main/java/org/jabref/model/openoffice/ootext/OOTextIntoOO.java

```diff
@@ -0,0 +1,817 @@
+        Pattern.compile("<(/" + TAG_NAME_REGEXP + ")>"
+                        + "|"
+                        + "<(" + TAG_NAME_REGEXP + ")"
+                        + "((?:\\s+(" + ATTRIBUTE_NAME_REGEXP + ")=" + ATTRIBUTE_VALUE_REGEXP + ")*)"
+                        + ">");
+
+    private static final Pattern ATTRIBUTE_PATTERN =
+        Pattern.compile("\\s+(" + ATTRIBUTE_NAME_REGEXP + ")=" + ATTRIBUTE_VALUE_REGEXP);
+
+    private OOTextIntoOO() {
+        // Hide the public constructor
+    }
+
+    /**
+     * Insert a text with formatting indicated by HTML-like tags, into
+     * a text at the position given by a cursor.
+     *
+     * Limitation: understands no entities. It does not receive any either, unless
+     * the user provides it.
+     *
+     * To limit the damage {@code TAG_NAME_REGEXP} and {@code ATTRIBUTE_NAME_REGEXP}
+     * explicitly lists the names we care about.
+     *
+     * Notable changes w.r.t insertOOFormattedTextAtCurrentLocation:
+     *
+     * - new tags:
+     *
+     *   - &lt;span lang="zxx"&gt;
+     *     - earlier was applied from code
+     *
+     *   - &lt;span oo:CharStyleName="CharStylename"&gt;
+     *     - earlier was applied from code, for "CitationCharacterFormat"
+     *
+     *   - &lt;p&gt; start new paragraph
+     *     - earlier was applied from code
+     *
+     *   - &lt;p oo:ParaStyleName="ParStyleName"&gt; : start new paragraph and apply ParStyleName
+     *     - earlier was applied from code
+     *
+     *   - &lt;tt&gt;
+     *     - earlier: known, but ignored
+     *     - now: equivalent to &lt;span oo:CharStyleName="Example"&gt;
+     *   - &lt;oo:referenceToPageNumberOfReferenceMark&gt; (self-closing)
+     *
+     * - closing tags try to properly restore state (in particular, the "not directly set" state)
+     *   instead of dictating an "off" state. This makes a difference when the value inherited from
+     *   another level (for example the paragraph) is not the "off" state.
+     *
+     *   An example: a style with
+     *   ReferenceParagraphFormat="JR_bibentry"
+     *   Assume JR_bibentry in LibreOffice is a paragraph style that prescribes "bold" font.
+     *   LAYOUT only prescribes bold around year.
+     *   Which parts of the bibliography entries should come out as bold?
+     *
+     * - The user can format citation marks (it is enough to format their start) and the
+     *   properties not (everywhere) dictated by the style are preserved (where they are not).
+     *
+     * @param position   The cursor giving the insert location. Not modified.
+     * @param ootext     The marked-up text to insert.
+     */
+    public static void write(XTextDocument doc, XTextCursor position, OOText ootext)
+        throws
+        UnknownPropertyException,
+        PropertyVetoException,
+        WrappedTargetException,
+        IllegalArgumentException,
+        NoSuchElementException,
+        CreationException {
+
+        final boolean debugThisFun = false;
```

Comment:
can this be removed? Better use LOGGER.debug(...)

Reply:
Changed to use LOGGER.debug() and not to use debugThisFun

- I believe LOGGER calls will be included in production code.
- Is this intended?

Reply:
Yes, this is fine. debug calls are only enabled with the debug flag starting parameter.